stage1/prepare-app/prepare-app.c

```diff
@@ -51,38 +51,94 @@ typedef struct _mount_point_t {
 #define dir(_name, _mode) \
 	{ .name = _name, .mode = _mode }
 
+int getmachinename(char *machine) {
+	static const int machineid_size = 32; // 0123456789abcdef0123456789ab
+	static const int machinename_size = machineid_size + 8; // rkt-01234567-89ab-cdef-0123-456789ab
+	int fd;
+	char buf[machineid_size + 1];
+	ssize_t r;
+
+	if ((fd = open("/etc/machine-id", O_RDONLY)) == -1)
+		return -1;
+
+	if ((r = read(fd, buf, machineid_size)) != machineid_size) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if ((machine = malloc(machinename_size + 1)) == NULL)
+		return -1;
+
+	if (snprintf(machine, machinename_size, "rkt-%.8s-%.4s-%.4s-%.4s-%.8s",
+		buf, buf+8, buf+12, buf+16, buf+20) == -1) {
+		return -1;
+	}
+	machine[machinename_size] = '\0';
+
+	return 0;
+}
+
+int writehosts(int rootfd) {
+	char buf[128];
+	int hostsfd;
+	char *machine;
+
+	if (getmachinename(machine) != 0)
+		goto fail;
+
+	if ((hostsfd = openat(rootfd, "etc/hosts", O_RDWR|O_CREAT, 0644)) < 0)
+		goto fail;
+
+	if (snprintf(buf, sizeof(buf), "%s\t%s\t%s\t%s\n", "127.0.0.1",
+			machine, "localhost", "localhost.localdomain") == -1)
+		goto fail;
+
+	if (write(hostsfd, buf, strlen(buf)) != strlen(buf))
+		goto fail;
+
+	if (close(hostsfd) != 0)
+		goto fail;
+
+	return 0;
+
+fail:
+	free(machine);
```

Comment:
> freeing machine but not closing hostsfd in the fail path?

Ooops, I forgot about that.

> in a trivial small program like this where any error in this function results in the program exiting I'd just replace all these if ( ) goto blocks with exit_if() and don't bother having a return code from writehosts(), and don't bother with cleanup paths.

OK. Since we exit on fail I guess ignoring cleanups is fine